src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java

```diff
@@ -431,26 +425,43 @@ private void beforeSlotChangedCraftingGrid(final InventoryCraftResult output, fi
         final ItemStackSnapshot repl = ItemStackUtil.snapshotOf(output.getStackInSlot(index));
 
         final SlotAdapter slot = this.impl$getAdapters().get(index);
-        this.impl$capturedCraftPreviewTransactions.add(new SlotTransaction(slot, orig, repl));
+        if (this.impl$capturedCraftPreviewTransaction == null) {
+            this.impl$capturedCraftPreviewTransaction = new SlotTransaction(slot, orig, repl);
+        } else if (this.impl$capturedCraftPreviewTransaction.getSlot().equals(slot)
+                && this.impl$capturedCraftPreviewTransaction.getFinal().equals(orig)) {
```

Comment:
Are you doing the correct equality check here? Or are you aiming to just allow instance equality checking the `ItemStackSnapshot`?

Reply:
It has to be `equals()` - `orig` is created fresh each time the method is called.